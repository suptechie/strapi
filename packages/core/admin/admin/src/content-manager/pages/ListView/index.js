import * as React from 'react';

import {
  IconButton,
  Main,
  Box,
  ActionLayout,
  Button,
  ContentLayout,
  HeaderLayout,
  useNotifyAT,
  Flex,
} from '@strapi/design-system';
import {
  NoPermissions,
  CheckPermissions,
  SearchURLQuery,
  useFetchClient,
  useFocusWhenNavigate,
  useQueryParams,
  useNotification,
  useRBACProvider,
  useTracking,
  Link,
  useAPIErrorHandler,
  getYupInnerErrors,
  useStrapiApp,
  DynamicTable,
  PaginationURLQuery,
  PageSizeURLQuery,
} from '@strapi/helper-plugin';
import { ArrowLeft, Cog, Plus } from '@strapi/icons';
import axios from 'axios';
import isEqual from 'lodash/isEqual';
import PropTypes from 'prop-types';
import { stringify } from 'qs';
import { useIntl } from 'react-intl';
import { useMutation } from 'react-query';
import { connect, useSelector } from 'react-redux';
import { useHistory, useLocation, Link as ReactRouterLink } from 'react-router-dom';
import { bindActionCreators, compose } from 'redux';
import styled from 'styled-components';

import { INJECT_COLUMN_IN_TABLE } from '../../../exposedHooks';
import { useEnterprise } from '../../../hooks/useEnterprise';
import { selectAdminPermissions } from '../../../pages/App/selectors';
import { InjectionZone } from '../../../shared/components';
import AttributeFilter from '../../components/AttributeFilter';
import BulkActionsBar from '../../components/DynamicTable/BulkActionsBar';
import { createYupSchema, getRequestUrl, getTrad } from '../../utils';

import { getData, getDataSucceeded, onChangeListHeaders, onResetListHeaders } from './actions';
import { ConfirmDialogDelete } from './components/ConfirmDialogDelete';
import { ConfirmDialogDeleteAll } from './components/ConfirmDialogDeleteAll';
import { FieldPicker } from './components/FieldPicker';
import { TableRows } from './components/TableRows';
import makeSelectListView, { selectDisplayedHeaders } from './selectors';
import { buildValidGetParams } from './utils';

const ConfigureLayoutBox = styled(Box)`
  svg {
    path {
      fill: ${({ theme }) => theme.colors.neutral900};
    }
  }
`;

const REVIEW_WORKFLOW_COLUMNS_CE = null;

function ListView({
  canCreate,
  canDelete,
  canRead,
  canPublish,
  data,
  getData,
  getDataSucceeded,
  isLoading,
  layout,
  pagination,
  slug,
}) {
  const { total } = pagination;
  const { contentType } = layout;
  const {
    info,
    options,
    metadatas,
    settings: { bulkable: isBulkable, filterable: isFilterable, searchable: isSearchable },
  } = contentType;

  const toggleNotification = useNotification();
  const { trackUsage } = useTracking();
  const { refetchPermissions } = useRBACProvider();
  const trackUsageRef = React.useRef(trackUsage);
  const fetchPermissionsRef = React.useRef(refetchPermissions);
  const { notifyStatus } = useNotifyAT();
  const { formatAPIError } = useAPIErrorHandler(getTrad);
  const permissions = useSelector(selectAdminPermissions);

  useFocusWhenNavigate();

  const [{ query }] = useQueryParams();
  const params = React.useMemo(() => buildValidGetParams(query), [query]);
  const pluginsQueryParams = stringify({ plugins: query.plugins }, { encode: false });

  const { pathname } = useLocation();
  const { push } = useHistory();
  const { formatMessage } = useIntl();
  const fetchClient = useFetchClient();

  const hasDraftAndPublish = options?.draftAndPublish ?? false;
  const hasReviewWorkflows = options?.reviewWorkflows ?? false;

  const reviewWorkflowColumns = useEnterprise(
    REVIEW_WORKFLOW_COLUMNS_CE,
    async () =>
      (
        await import(
          '../../../../../ee/admin/content-manager/pages/ListView/ReviewWorkflowsColumn/constants'
        )
      ).REVIEW_WORKFLOW_COLUMNS_EE,
    {
      enabled: !!options?.reviewWorkflows,
    }
  );

  const { post, del } = fetchClient;

  const bulkPublishMutation = useMutation(
    (data) =>
      post(`/content-manager/collection-types/${contentType.uid}/actions/bulkPublish`, data),
    {
      onSuccess() {
        toggleNotification({
          type: 'success',
          message: { id: 'content-manager.success.record.publish', defaultMessage: 'Published' },
        });

        fetchData(`/content-manager/collection-types/${slug}`, { params });
      },
      onError(error) {
        toggleNotification({
          type: 'warning',
          message: formatAPIError(error),
        });
      },
    }
  );
  const bulkUnpublishMutation = useMutation(
    (data) =>
      post(`/content-manager/collection-types/${contentType.uid}/actions/bulkUnpublish`, data),
    {
      onSuccess() {
        toggleNotification({
          type: 'success',
          message: {
            id: 'content-manager.success.record.unpublish',
            defaultMessage: 'Unpublished',
          },
        });

        fetchData(`/content-manager/collection-types/${slug}`, { params });
      },
      onError(error) {
        toggleNotification({
          type: 'warning',
          message: formatAPIError(error),
        });
      },
    }
  );

  // FIXME
  // Using a ref to avoid requests being fired multiple times on slug on change
  // We need it because the hook as mulitple dependencies so it may run before the permissions have checked
  const requestUrlRef = React.useRef('');
  /**
   * TODO: re-write all of this, it's a mess.
   */
  const fetchData = React.useCallback(
    async (endPoint, options) => {
      getData();

      try {
        const {
          data: { results, pagination: paginationResult },
        } = await fetchClient.get(endPoint, options);

        notifyStatus(
          formatMessage(
            {
              id: getTrad('utils.data-loaded'),
              defaultMessage:
                '{number, plural, =1 {# entry has} other {# entries have}} successfully been loaded',
            },
            // Using the plural form
            { number: paginationResult.count }
          )
        );

        getDataSucceeded(paginationResult, results);
      } catch (err) {
        if (axios.isCancel(err)) {
          return;
        }

        const resStatus = err?.response?.status ?? null;

        if (resStatus === 403) {
          await fetchPermissionsRef.current();

          toggleNotification({
            type: 'info',
            message: { id: getTrad('permissions.not-allowed.update') },
          });

          push('/');

          return;
        }

        toggleNotification({
          type: 'warning',
          message: { id: getTrad('error.model.fetch') },
        });
      }
    },
    [formatMessage, getData, getDataSucceeded, notifyStatus, push, toggleNotification, fetchClient]
  );

  const handleConfirmDeleteAllData = React.useCallback(
    async (ids) => {
      try {
        await post(`/content-manager/collection-types/${slug}/actions/bulkDelete`, {
          ids,
        });

        fetchData(`/content-manager/collection-types/${slug}`, { params });

        trackUsageRef.current('didBulkDeleteEntries');
      } catch (err) {
        toggleNotification({
          type: 'warning',
          message: formatAPIError(err),
        });
      }
    },
    [slug, toggleNotification, formatAPIError, post, fetchData, params]
  );

  const handleConfirmDeleteData = React.useCallback(
    async (idToDelete) => {
      try {
        await del(`/content-manager/collection-types/${slug}/${idToDelete}`);

        const requestUrl = getRequestUrl(`collection-types/${slug}`);
        fetchData(requestUrl, { params });

        toggleNotification({
          type: 'success',
          message: { id: getTrad('success.record.delete') },
        });
      } catch (err) {
        toggleNotification({
          type: 'warning',
          message: formatAPIError(err),
        });
      }
    },
    [slug, toggleNotification, formatAPIError, del, fetchData, params]
  );

  /**
   * @param {number[]} selectedEntries - Array of ids to publish
   * @returns {{validIds: number[], errors: Object.<number, string>}} - Returns an object with the valid ids and the errors
   */
  const validateEntriesToPublish = async (selectedEntries) => {
    const validations = { validIds: [], errors: {} };
    // Create the validation schema based on the contentType
    const schema = createYupSchema(
      contentType,
      { components: layout.components },
      { isDraft: false }
    );
    // Get the selected entries
    const entries = data.filter((entry) => {
      return selectedEntries.includes(entry.id);
    });
    // Validate each entry and map the unresolved promises
    const validationPromises = entries.map((entry) =>
      schema.validate(entry, { abortEarly: false })
    );
    // Resolve all the promises in one go
    const resolvedPromises = await Promise.allSettled(validationPromises);
    // Set the validations
    resolvedPromises.forEach((promise) => {
      if (promise.status === 'rejected') {
        const entityId = promise.reason.value.id;
        validations.errors[entityId] = getYupInnerErrors(promise.reason);
      }

      if (promise.status === 'fulfilled') {
        validations.validIds.push(promise.value.id);
      }
    });

    return validations;
  };

  const handleConfirmPublishAllData = async (selectedEntries) => {
    const validations = await validateEntriesToPublish(selectedEntries);

    if (Object.values(validations.errors).length) {
      toggleNotification({
        type: 'warning',
        title: {
          id: 'content-manager.listView.validation.errors.title',
          defaultMessage: 'Action required',
        },
        message: {
          id: 'content-manager.listView.validation.errors.message',
          defaultMessage:
            'Please make sure all fields are valid before publishing (required field, min/max character limit, etc.)',
        },
      });

      throw new Error('Validation error');
    }

    return bulkPublishMutation.mutateAsync({ ids: selectedEntries });
  };

  const handleConfirmUnpublishAllData = (selectedEntries) => {
    return bulkUnpublishMutation.mutateAsync({ ids: selectedEntries });
  };

  React.useEffect(() => {
    const CancelToken = axios.CancelToken;
    const source = CancelToken.source();

    const shouldSendRequest = canRead;
    const requestUrl = getRequestUrl(`collection-types/${slug}`);

    if (shouldSendRequest && requestUrl.includes(requestUrlRef.current)) {
      fetchData(requestUrl, { cancelToken: source.token, params });
    }

    return () => {
      requestUrlRef.current = slug;

      source.cancel('Operation canceled by the user.');
    };
  }, [canRead, getData, slug, params, getDataSucceeded, fetchData]);

  const defaultHeaderLayoutTitle = formatMessage({
    id: getTrad('header.name'),
    defaultMessage: 'Content',
  });
  const headerLayoutTitle = formatMessage({
    id: info.displayName,
    defaultMessage: info.displayName || defaultHeaderLayoutTitle,
  });

  const { runHookWaterfall } = useStrapiApp();
  const displayedHeaders = useSelector(selectDisplayedHeaders);

  const tableHeaders = React.useMemo(() => {
    const headers = runHookWaterfall(INJECT_COLUMN_IN_TABLE, {
      displayedHeaders,
      layout,
    });

    const formattedHeaders = headers.displayedHeaders.map((header) => {
      const { metadatas } = header;

      if (header.fieldSchema.type === 'relation') {
        const sortFieldValue = `${header.name}.${header.metadatas.mainField.name}`;

        return {
          ...header,
          metadatas: {
            ...metadatas,
            label: formatMessage({
              id: getTrad(`containers.ListPage.table-headers.${header.name}`),
              defaultMessage: metadatas.label,
            }),
          },
          name: sortFieldValue,
        };
      }

      return {
        ...header,
        metadatas: {
          ...metadatas,
          label: formatMessage({
            id: getTrad(`containers.ListPage.table-headers.${header.name}`),
            defaultMessage: metadatas.label,
          }),
        },
      };
    });

    if (hasDraftAndPublish) {
      formattedHeaders.push({
        key: '__published_at_temp_key__',
        name: 'publishedAt',
        fieldSchema: {
          type: 'custom',
        },
        metadatas: {
          label: formatMessage({
            id: getTrad(`containers.ListPage.table-headers.publishedAt`),
            defaultMessage: 'publishedAt',
          }),
          searchable: false,
          sortable: true,
        },
      });
    }

    if (reviewWorkflowColumns) {
      // Make sure the column header label is translated
      if (typeof reviewWorkflowColumns.metadatas.label !== 'string') {
        reviewWorkflowColumns.metadatas.label = formatMessage(
          reviewWorkflowColumns.metadatas.label
        );
      }

      formattedHeaders.push(reviewWorkflowColumns);
    }

    return formattedHeaders;
  }, [
    runHookWaterfall,
    displayedHeaders,
    layout,
    reviewWorkflowColumns,
    hasDraftAndPublish,
    formatMessage,
  ]);

  const subtitle = canRead
    ? formatMessage(
        {
          id: getTrad('pages.ListView.header-subtitle'),
          defaultMessage: '{number, plural, =0 {# entries} one {# entry} other {# entries}} found',
        },
        { number: total }
      )
    : null;

  const getCreateAction = (props) =>
    canCreate ? (
      <Button
        {...props}
        forwardedAs={ReactRouterLink}
        onClick={() => {
          const trackerProperty = hasDraftAndPublish ? { status: 'draft' } : {};

          trackUsageRef.current('willCreateEntry', trackerProperty);
        }}
        to={{
          pathname: `${pathname}/create`,
          search: query.plugins ? pluginsQueryParams : '',
        }}
        startIcon={<Plus />}
        style={{ textDecoration: 'none' }}
      >
        {formatMessage({
          id: getTrad('HeaderLayout.button.label-add-entry'),
          defaultMessage: 'Create new entry',
        })}
      </Button>
    ) : null;

  return (
    <Main aria-busy={isLoading}>
      <HeaderLayout
        primaryAction={getCreateAction()}
        subtitle={subtitle}
        title={headerLayoutTitle}
        navigationAction={
          <Link startIcon={<ArrowLeft />} to="/content-manager/">
            {formatMessage({
              id: 'global.back',
              defaultMessage: 'Back',
            })}
          </Link>
        }
      />
      {!canRead && (
        <ActionLayout endActions={<InjectionZone area="contentManager.listView.actions" />} />
      )}
      {canRead && (
        <ActionLayout
          endActions={
            <>
              <InjectionZone area="contentManager.listView.actions" />
              <FieldPicker layout={layout} />
              <CheckPermissions
                permissions={permissions.contentManager.collectionTypesConfigurations}
              >
                <ConfigureLayoutBox paddingTop={1} paddingBottom={1}>
                  <IconButton
                    onClick={() => {
                      trackUsage('willEditListLayout');
                    }}
                    forwardedAs={ReactRouterLink}
                    to={{ pathname: `${slug}/configurations/list`, search: pluginsQueryParams }}
                    icon={<Cog />}
                    label={formatMessage({
                      id: 'app.links.configure-view',
                      defaultMessage: 'Configure the view',
                    })}
                  />
                </ConfigureLayoutBox>
              </CheckPermissions>
            </>
          }
          startActions={
            <>
              {isSearchable && (
                <SearchURLQuery
                  label={formatMessage(
                    { id: 'app.component.search.label', defaultMessage: 'Search for {target}' },
                    { target: headerLayoutTitle }
                  )}
                  placeholder={formatMessage({
                    id: 'global.search',
                    defaultMessage: 'Search',
                  })}
                  trackedEvent="didSearch"
                />
              )}
              {isFilterable && (
                <AttributeFilter contentType={contentType} slug={slug} metadatas={metadatas} />
              )}
            </>
          }
        />
      )}
      <ContentLayout>
        {canRead ? (
          <Flex gap={4} direction="column" alignItems="stretch">
            <DynamicTable
              canCreate={canCreate}
              canDelete={canDelete}
              canPublish={canPublish}
              contentTypeName={headerLayoutTitle}
              onConfirmDelete={handleConfirmDeleteData}
              onConfirmDeleteAll={handleConfirmDeleteAllData}
              onConfirmPublishAll={handleConfirmPublishAllData}
              onConfirmUnpublishAll={handleConfirmUnpublishAllData}
              isBulkable={isBulkable}
              isLoading={isLoading}
              // FIXME: remove the layout props drilling
              layout={layout}
              rows={data}
              components={{ ConfirmDialogDelete, ConfirmDialogDeleteAll }}
              contentType={headerLayoutTitle}
              action={getCreateAction({ variant: 'secondary' })}
              headers={tableHeaders}
              onOpenDeleteAllModalTrackedEvent="willBulkDeleteEntries"
              withBulkActions
              withMainAction={canDelete && isBulkable}
              renderBulkActionsBar={({ selectedEntries, clearSelectedEntries }) => (
                <BulkActionsBar
                  showPublish={canPublish && hasDraftAndPublish}
                  showDelete={canDelete}
                  onConfirmDeleteAll={handleConfirmDeleteAllData}
                  onConfirmPublishAll={handleConfirmPublishAllData}
                  onConfirmUnpublishAll={handleConfirmUnpublishAllData}
                  selectedEntries={selectedEntries}
                  clearSelectedEntries={clearSelectedEntries}
                />
              )}
              bulkAction
            >
              <TableRows
                canCreate={canCreate}
                canDelete={canDelete}
                contentType={contentType}
                features={{
                  hasDraftAndPublish,
                  hasReviewWorkflows,
                }}
                headers={tableHeaders}
                rows={data}
                withBulkActions
                withMainAction={canDelete && isBulkable}
              />
            </DynamicTable>
            <Flex alignItems="flex-end" justifyContent="space-between">
              <PageSizeURLQuery trackedEvent="willChangeNumberOfEntriesPerPage" />
              <PaginationURLQuery pagination={{ pageCount: pagination?.pageCount || 1 }} />
            </Flex>
          </Flex>
        ) : (
          <NoPermissions />
        )}
      </ContentLayout>
    </Main>
  );
}

ListView.propTypes = {
  canCreate: PropTypes.bool.isRequired,
  canDelete: PropTypes.bool.isRequired,
  canRead: PropTypes.bool.isRequired,
  canPublish: PropTypes.bool.isRequired,
  data: PropTypes.array.isRequired,
  layout: PropTypes.exact({
    components: PropTypes.object.isRequired,
    contentType: PropTypes.shape({
      uid: PropTypes.string.isRequired,
      attributes: PropTypes.object.isRequired,
      metadatas: PropTypes.object.isRequired,
      info: PropTypes.shape({ displayName: PropTypes.string.isRequired }).isRequired,
      layouts: PropTypes.shape({
        list: PropTypes.array.isRequired,
      }).isRequired,
      options: PropTypes.object.isRequired,
      settings: PropTypes.object.isRequired,
    }).isRequired,
  }).isRequired,
  isLoading: PropTypes.bool.isRequired,
  getData: PropTypes.func.isRequired,
  getDataSucceeded: PropTypes.func.isRequired,
  pagination: PropTypes.shape({ total: PropTypes.number.isRequired, pageCount: PropTypes.number })
    .isRequired,
  slug: PropTypes.string.isRequired,
};

const mapStateToProps = makeSelectListView();

export function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      getData,
      getDataSucceeded,
      onChangeListHeaders,
      onResetListHeaders,
    },
    dispatch
  );
}
const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect)(React.memo(ListView, isEqual));
