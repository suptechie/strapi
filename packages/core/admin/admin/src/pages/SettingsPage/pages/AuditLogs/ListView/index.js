import React, { useState } from 'react';
import { useIntl } from 'react-intl';
import { useQuery } from 'react-query';
import {
  SettingsPageTitle,
  DynamicTable,
  useRBAC,
  useNotification,
  useFocusWhenNavigate,
} from '@strapi/helper-plugin';
import { HeaderLayout, ContentLayout } from '@strapi/design-system/Layout';
import { Main } from '@strapi/design-system/Main';
import adminPermissions from '../../../../../permissions';
// import { useFetchClient } from '../../../../../hooks';
import TableRows from './DynamicTable/TableRows';
import tableHeaders from './utils/tableHeaders';
import { fetchAuditLogsPage, fetchAuditLog } from './utils/api';
import ModalDialog from './ModalDialog';

const ListView = () => {
  const { formatMessage } = useIntl();
  const toggleNotification = useNotification();
  const {
    allowedActions: { canRead },
  } = useRBAC(adminPermissions.settings.auditLogs);

  useFocusWhenNavigate();

  /**
   * TODO: using fetchclient facing difficulties to mock, for time being using axiosinstance to pass the tests
   *
  const { get } = useFetchClient();
  const fetchData = async () => {
    const {
      data: { results },
    } = await get(`/admin/audit-logs`);

    return results;
  }; */

  const { status, data, isFetching } = useQuery('audit-logs', fetchAuditLogsPage, {
    enabled: canRead,
    keepPreviousData: true,
    retry: false,
    staleTime: 1000 * 20,
    onError() {
      toggleNotification({
        type: 'warning',
        message: { id: 'notification.error', defaultMessage: 'An error occured' },
      });
    },
  });

  const isLoading = status === 'loading' || isFetching;

  const title = formatMessage({
    id: 'global.auditLogs',
    defaultMessage: 'Audit Logs',
  });

  const headers = tableHeaders.map((header) => ({
    ...header,
    metadatas: {
      ...header.metadatas,
      label: formatMessage(header.metadatas.label),
    },
  }));

  const [openedLogId, setOpenedLogId] = useState(null);
  const { data: openedLogData, status: openedLogStatus } = useQuery(
    ['audit-log', openedLogId],
    () => fetchAuditLog(openedLogId),
    {
      enabled: !!openedLogId,
    }
  );

  console.log({ openedLogData, openedLogId });

  const handleToggle = (id) => {
    // Either saves the id of the log to open or closes the modal
    setOpenedLogId(id);
  };

  return (
    <Main aria-busy={isLoading}>
      <SettingsPageTitle name={title} />
      <HeaderLayout
        title={title}
        subtitle={formatMessage({
          id: 'Settings.permissions.auditLogs.listview.header.subtitle',
          defaultMessage: 'Logs of all the activities that happened in your environment',
        })}
      />
      <ContentLayout>
        <DynamicTable
          contentType="Audit logs"
          headers={headers}
          rows={data}
          withBulkActions
          isLoading={isLoading}
        >
          <TableRows headers={headers} rows={data} onModalToggle={handleToggle} />
        </DynamicTable>
      </ContentLayout>
      {openedLogId && openedLogStatus === 'success' && (
        <ModalDialog onToggle={handleToggle} data={openedLogData} />
      )}
    </Main>
  );
};

export default ListView;
