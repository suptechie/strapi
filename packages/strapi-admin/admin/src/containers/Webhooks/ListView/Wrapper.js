/**
 *
 * Wrapper
 *
 */

import styled from 'styled-components';

const Wrapper = styled.div`
  > div:first-of-type {
    margin-bottom: 33px;
  }
  .list-wrapper {
    border-radius: 3px;
    box-shadow: 0 2px 4px #e3e9f3;
    background: white;
    > div,
    > div > div:last-of-type {
      box-shadow: none;
      border-bottom-left-radius: 0;
      border-bottom-right-radius: 0;
    }
  }
  p {
    margin-bottom: 0;
  }
`;

export default Wrapper;
