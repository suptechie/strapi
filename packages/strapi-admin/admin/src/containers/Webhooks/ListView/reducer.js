import { fromJS } from 'immutable';

const initialState = fromJS({
  webhooks: [],
});

const reducer = (state, action) => {
  switch (action.type) {
    case 'GET_DATA_SUCCEEDED':
      return state.update('webhooks', () => fromJS(action.data));
    case 'SET_WEBHOOK_ENABLED':
      return state.updateIn(['webhooks', ...action.keys], () => action.value);
    default:
      return state;
  }
};

export default reducer;
export { initialState };
