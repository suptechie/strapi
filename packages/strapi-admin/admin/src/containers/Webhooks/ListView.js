/**
 *
 * ListView
 *
 */

import React, { memo } from 'react';

import { Header } from '@buffetjs/custom';
import { useGlobalContext } from 'strapi-helper-plugin';
import { List } from '@buffetjs/custom';

import ListRow from '../../components/ListRow';
import Wrapper from './Wrapper';

import useDataManager from '../../hooks/useDataManager';

function ListView() {
  const { formatMessage } = useGlobalContext();
  const { webhooks } = useDataManager();

  // Header props
  const actions = [
    {
      title: formatMessage({ id: `Settings.webhook.list.button.add` }),
      onClick: () => {},
      color: 'primary',
      type: 'button',
      icon: true,
      style: {
        paddingLeft: 15,
        paddingRight: 15,
      },
    },
  ];

  const headerProps = {
    title: {
      label: formatMessage({ id: `Settings.webhook.list.title` }),
    },
    content: formatMessage({ id: `Settings.webhook.list.description` }),
    actions: actions,
  };

  // List props
  const rowsCount = webhooks.length;
  const titleLabel = `${
    rowsCount > 1
      ? formatMessage({ id: `Settings.webhook.list.label.plur` })
      : formatMessage({ id: `Settings.webhook.list.label.sing` })
  }`;
  const title = `${rowsCount} ${titleLabel}`;
  const buttonProps = {
    color: 'secondary',
    label: formatMessage({ id: `Settings.webhook.list.button.delete` }),
    onClick: () => {},
    type: 'button',
  };
  const listProps = {
    title,
    button: buttonProps,
    items: webhooks,
  };

  return (
    <Wrapper>
      <Header {...headerProps} />
      <List
        {...listProps}
        customRowComponent={props => <ListRow {...props} />}
      />
    </Wrapper>
  );
}

export default memo(ListView);
