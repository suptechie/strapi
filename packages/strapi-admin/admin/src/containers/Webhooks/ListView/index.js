/**
 *
 * ListView
 *
 */

import React, { useEffect, useReducer, useState } from 'react';
import { useHistory, useLocation } from 'react-router-dom';

import { Header, List } from '@buffetjs/custom';
import { Button } from '@buffetjs/core';
import { Plus } from '@buffetjs/icons';

import {
  request,
  useGlobalContext,
  ListButton,
  PopUpWarning,
} from 'strapi-helper-plugin';

import ListRow from '../../../components/ListRow';
import EmptyList from '../../../components/EmptyList';
import Wrapper from './Wrapper';

import reducer, { initialState } from './reducer';

function ListView() {
  const { formatMessage } = useGlobalContext();
  const [showModal, setShowModal] = useState(false);
  const [reducerState, dispatch] = useReducer(reducer, initialState);
  const { push } = useHistory();
  const { pathname } = useLocation();

  const { webhooks, webhooksToDelete } = reducerState.toJS();

  useEffect(() => {
    fetchData();
  }, []);

  const getWebhookIndex = id =>
    webhooks.findIndex(webhook => webhook.id === id);

  // New button
  const addBtnLabel = formatMessage({
    id: `Settings.webhooks.list.button.add`,
  });

  const newButtonProps = {
    label: addBtnLabel,
    onClick: () => handleGoTo('create'),
    color: 'primary',
    type: 'button',
    icon: <Plus fill="#007eff" width="11px" height="11px" />,
  };

  // Header props
  const actions = [
    {
      ...newButtonProps,
      icon: true,
      style: {
        paddingLeft: 15,
        paddingRight: 15,
      },
    },
  ];

  const headerProps = {
    title: {
      label: formatMessage({ id: `Settings.webhooks.title` }),
    },
    content: formatMessage({ id: `Settings.webhooks.list.description` }),
    actions: actions,
  };

  // List props
  const rowsCount = webhooks.length;
  const titleLabel = `${
    rowsCount > 1
      ? formatMessage({ id: `Settings.webhooks.title` })
      : formatMessage({ id: `Settings.webhooks.singular` })
  }`;
  const title = `${rowsCount} ${titleLabel}`;

  const buttonProps = {
    color: 'delete',
    disabled: webhooksToDelete.length > 0 ? false : true,
    label: formatMessage({ id: `Settings.webhooks.list.button.delete` }),
    onClick: () => setShowModal(true),
    type: 'button',
  };

  const listProps = {
    title,
    button: buttonProps,
    items: webhooks,
  };

  const fetchData = async () => {
    try {
      const { data } = await request(`/admin/webhooks`, {
        method: 'GET',
      });

      dispatch({
        type: 'GET_DATA_SUCCEEDED',
        data,
      });
    } catch (err) {
      if (err.code !== 20) {
        strapi.notification.error('notification.error');
      }
    }
  };

  const handleChange = (value, id) => {
    const updatedWebhooksToDelete = value
      ? [...webhooksToDelete, id]
      : webhooksToDelete.filter(webhookId => webhookId !== id);

    dispatch({
      type: 'SET_WEBHOOKS_TO_DELETE',
      webhooks: updatedWebhooksToDelete,
    });
  };

  const handleDeleteAllConfirm = async () => {
    await onDeleteAllCLick();
    setShowModal(false);
  };

  const handleDeleteConfirm = async id => {
    await onDeleteCLick(id);
    setShowModal(false);
  };

  const handleEnabledChange = async (value, id) => {
    const webhookIndex = getWebhookIndex(id);

    const initialWebhookProps = webhooks[webhookIndex];
    const keys = [webhookIndex, 'isEnabled'];

    const body = {
      ...initialWebhookProps,
      isEnabled: value,
    };

    delete body.id;

    try {
      dispatch({
        type: 'SET_WEBHOOK_ENABLED',
        keys,
        value: value,
      });

      await request(`/admin/webhooks/${id}`, {
        method: 'PUT',
        body,
      });
    } catch (err) {
      dispatch({
        type: 'SET_WEBHOOK_ENABLED',
        keys,
        value: !value,
      });

      if (err.code !== 20) {
        strapi.notification.error('notification.error');
      }
    }
  };

  const handleGoTo = to => {
    push(`${pathname}/${to}`);
  };

  const onDeleteAllCLick = async () => {
    const body = {
      ids: webhooksToDelete,
    };

    try {
      await request(`/admin/webhooks/batch-delete`, {
        method: 'POST',
        body,
      });

      dispatch({
        type: 'WEBHOOKS_DELETED',
      });
    } catch (err) {
      if (err.code !== 20) {
        strapi.notification.error('notification.error');
      }
    }
  };

  const onDeleteCLick = async id => {
    try {
      await request(`/admin/webhooks/${id}`, {
        method: 'DELETE',
      });

      dispatch({
        type: 'WEBHOOK_DELETED',
        index: getWebhookIndex(id),
      });
    } catch (err) {
      if (err.code !== 20) {
        strapi.notification.error('notification.error');
      }
    }
  };

  return (
    <Wrapper>
      <Header {...headerProps} />
      <div className="list-wrapper">
        {rowsCount > 0 ? (
          <List
            {...listProps}
            customRowComponent={props => {
              return (
                <ListRow
                  {...props}
                  onCheckChange={handleChange}
                  onEditClick={handleGoTo}
                  onDeleteCLick={handleDeleteConfirm}
                  onEnabledChange={handleEnabledChange}
                  itemsToDelete={webhooksToDelete}
                />
              );
            }}
          />
        ) : (
          <EmptyList />
        )}
        <ListButton>
          <Button {...newButtonProps} />
        </ListButton>
      </div>
      <PopUpWarning
        isOpen={showModal}
        toggleModal={() => setShowModal(!showModal)}
        popUpWarningType="danger"
        onConfirm={handleDeleteAllConfirm}
      />
    </Wrapper>
  );
}

export default ListView;
