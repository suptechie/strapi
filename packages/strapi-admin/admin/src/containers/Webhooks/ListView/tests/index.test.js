import React from 'react';
import { shallow } from 'enzyme';

import { GlobalContextProvider } from 'strapi-helper-plugin';
import ListView from '../index';

describe('Admin | containers | Webhooks | ListView | main', () => {
  describe('Render', () => {
    it('It should render properly', () => {
      const ListViewComponent = () => {
        return (
          <GlobalContextProvider>
            <ListView />
          </GlobalContextProvider>
        );
      };
      shallow(<ListViewComponent />);
    });
  });
});
