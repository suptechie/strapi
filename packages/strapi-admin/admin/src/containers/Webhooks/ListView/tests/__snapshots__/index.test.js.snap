// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin | containers | ListView should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bXGyLb lcTfTV"
  >
    <div
      class="sc-chPdSV jmcPdM"
    >
      <div
        class="sticky-wrapper"
      >
        <div
          class="row"
        >
          <div
            class="col-sm-6 header-title"
          >
            <div
              class="sc-kGXeez idElYQ"
            >
              <h1>
                Webhooks
              </h1>
            </div>
            <p>
              Get POST changes notifications.
            </p>
          </div>
          <div
            class="col-sm-6 justify-content-end"
          >
            <div
              class="sc-kgoBCf hhYBts"
            >
              <button
                class="sc-bxivhb cRlNra"
                color="primary"
                label="Add new webhook"
                style="padding-left: 15px; padding-right: 15px;"
                type="button"
              >
                <svg
                  fill="#fff"
                  height="11px"
                  viewBox="0 0 13 13"
                  width="11px"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g
                    data-name="Layer 2"
                  >
                    <path
                      d="M6.5.5v12m6-6H.5"
                      fill="none"
                      stroke="#fff"
                      stroke-linecap="round"
                    />
                  </g>
                </svg>
                Add new webhook
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="list-wrapper"
    >
      <div
        class="sc-daURTG jsRXtU"
      >
        <p>
          There are no webhooks yet
        </p>
        <p>
          Add your first one to this list.
        </p>
        <a
          href="https://strapi.io/documentation/3.0.0-beta.x/guides/webhooks.html"
          rel="noopener noreferrer"
          target="_blank"
        >
          See our documentation
        </a>
      </div>
      <div
        class="sc-gxMtzJ gUEbUR"
      >
        <button
          class="sc-bxivhb cRlNra"
          color="primary"
          label="Add new webhook"
          type="button"
        >
          <span
            class="sc-iRbamj bybvAC"
          >
            <svg
              fill="#007eff"
              height="11px"
              viewBox="0 0 13 13"
              width="11px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                data-name="Layer 2"
              >
                <path
                  d="M6.5.5v12m6-6H.5"
                  fill="none"
                  stroke="#007eff"
                  stroke-linecap="round"
                />
              </g>
            </svg>
          </span>
          Add new webhook
        </button>
      </div>
    </div>
    <div
      class="sc-hMFtBS eDSczM"
    />
  </div>
</DocumentFragment>
`;
